<?php

declare(strict_types=1);

namespace Accounting\Tests\Suppliers;

use PHPUnit\Framework\TestCase;
use Accounting\Database\DatabaseException;
use Accounting\Suppliers\{ SuppliersDto, ISuppliersRepository, SuppliersRepository };

class SuppliersRepositoryTest extends TestCase
{
    private $db;
    private $result;
    private array $input;
    private SuppliersDto $dto;
    private ISuppliersRepository $repository;

    protected function setUp(): void
    {
        $this->db = $this->createMock("Accounting\Database\IDatabase");
        $this->result = $this->createMock("Accounting\Database\IDatabaseResult");
        $this->input = [
            "id" => 6605,
            "name" => "Benefit system let if coach garden hair citizen.",
            "contact_person" => "Plan skill bad should visit. Already order rest process what. Investment school only able senior painting.",
            "email" => "robert26@example.com",
            "phone" => "Scientist employee hotel where speak. Back board subject recognize. Difficult summer road federal. Skill threat must decade.",
            "fax" => "Son south skin woman between ok. Rather pressure still economy life director.",
            "address" => "Fall bank boy hotel. System measure tax camera.",
        ];
        $this->dto = new SuppliersDto($this->input);
        $this->repository = new SuppliersRepository($this->db);
    }

    protected function tearDown(): void
    {
        unset($this->db);
        unset($this->result);
        unset($this->input);
        unset($this->dto);
        unset($this->repository);
    }

    public function testInsert_ReturnsFailedOnException(): void
    {
        $expected = -1;

        $this->db->method("prepare")
            ->will($this->throwException(new DatabaseException()));

        $actual = $this->repository->insert($this->dto);
        $this->assertEquals($expected, $actual);
    }

    public function testInsert_ReturnsId(): void
    {
        $expected = 9628;

        $sql = "INSERT INTO `Suppliers` (`name`, `contact_person`, `email`, `phone`, `fax`, `address`)
                VALUES (?, ?, ?, ?, ?, ?)";

        $this->db->expects($this->once())
            ->method("prepare")
            ->with($sql)
            ->willReturn($this->result);
        $this->result->expects($this->once())
            ->method("execute")
            ->with([
                $this->dto->name,
                $this->dto->contactPerson,
                $this->dto->email,
                $this->dto->phone,
                $this->dto->fax,
                $this->dto->address
            ]);
        $this->db->expects($this->once())
            ->method("lastInsertId")
            ->willReturn($expected);

        $actual = $this->repository->insert($this->dto);
        $this->assertEquals($expected, $actual);
    }

    public function testUpdate_ReturnsFailedOnException(): void
    {
        $expected = -1;

        $this->db->method("prepare")
            ->will($this->throwException(new DatabaseException()));

        $actual = $this->repository->update($this->dto);
        $this->assertEquals($expected, $actual);
    }

    public function testUpdate_ReturnsRowCount(): void
    {
        $expected = 7862;

        $sql = "UPDATE `Suppliers` SET `name` = ?, `contact_person` = ?, `email` = ?, `phone` = ?, `fax` = ?, `address` = ?
                WHERE `id` = ?";

        $this->db->expects($this->once())
            ->method("prepare")
            ->with($sql)
            ->willReturn($this->result);
        $this->result->expects($this->once())
            ->method("execute")
            ->with([
                $this->dto->name,
                $this->dto->contactPerson,
                $this->dto->email,
                $this->dto->phone,
                $this->dto->fax,
                $this->dto->address,
                $this->dto->id
            ]);
        $this->result->expects($this->once())
            ->method("rowCount")
            ->willReturn($expected);

        $actual = $this->repository->update($this->dto);
        $this->assertEquals($expected, $actual);
    }

    public function testGet_ReturnsEmptyOnException(): void
    {
        $id = 2758;

        $this->db->method("prepare")
            ->will($this->throwException(new DatabaseException()));

        $actual = $this->repository->get($id);
        $this->assertEmpty($actual);
    }

    public function testGet_ReturnsDto(): void
    {
        $id = 4765;

        $sql = "SELECT `id`, `name`, `contact_person`, `email`, `phone`, `fax`, `address`
                FROM `Suppliers` WHERE `id` = ?";

        $this->db->expects($this->once())
            ->method("prepare")
            ->with($sql)
            ->willReturn($this->result);
        $this->result->expects($this->once())
            ->method("execute")
            ->with([$id]);
        $this->result->expects($this->once())
            ->method("fetchAll")
            ->willReturn([$this->input]);

        $actual = $this->repository->get($id);
        $this->assertEquals($this->dto, $actual);
    }

    public function testGetAll_ReturnsEmptyOnException(): void
    {
        $this->db->method("prepare")
            ->will($this->throwException(new DatabaseException()));

        $actual = $this->repository->getAll();
        $this->assertEmpty($actual);
    }

    public function testGetAll_ReturnsDtos(): void
    {
        $sql = "SELECT `id`, `name`, `contact_person`, `email`, `phone`, `fax`, `address`
                FROM `Suppliers`";

        $this->db->expects($this->once())
            ->method("prepare")
            ->with($sql)
            ->willReturn($this->result);
        $this->result->expects($this->once())
            ->method("execute");
        $this->result->expects($this->once())
            ->method("fetchAll")
            ->willReturn([$this->input]);

        $actual = $this->repository->getAll();
        $this->assertEquals([$this->dto], $actual);
    }

    public function testDelete_ReturnsFailedOnException(): void
    {
        $id = 4320;
        $expected = -1;

        $this->db->method("prepare")
            ->will($this->throwException(new DatabaseException()));

        $actual = $this->repository->delete($id);
        $this->assertEquals($expected, $actual);
    }

    public function testDelete_ReturnsRowCount(): void
    {
        $id = 8300;
        $expected = 3552;

        $sql = "DELETE FROM `Suppliers` WHERE `id` = ?";

        $this->db->expects($this->once())
            ->method("prepare")
            ->with($sql)
            ->willReturn($this->result);
        $this->result->expects($this->once())
            ->method("execute")
            ->with([$id]);
        $this->result->expects($this->once())
            ->method("rowCount")
            ->willReturn($expected);

        $actual = $this->repository->delete($id);
        $this->assertEquals($expected, $actual);
    }
}